src/coreclr/jit/lclmorph.cpp

```diff
@@ -2426,3 +2426,562 @@ void Compiler::fgExposeLocalsInBitVec(BitVec_ValArg_T bitVec)
+}
+
+//-----------------------------------------------------------------------------
+// fgLCLMasksUpdateLCLStore: For the given statement, if it is a local store,
+// and the weighting recommends to switch, then update to store as a mask.
+//
+// Arguments:
+//     stmt - The statement.
+//     weightsTable - table to update.
+//
+// Returns:
+//     True if a converted local store was found.
+//
+bool Compiler::fgLCLMasksUpdateLCLStore(Statement* stmt, LCLMasksWeightTable* weightsTable)
+{
+    // Look for:
+    //      STORELCL(TYP_SIMD, ConvertMaskToVector(mask))
+
+    GenTree* tree = stmt->GetRootNode();
+
+    if ((tree->OperGet() != GT_STORE_LCL_VAR) || (!varTypeIsSIMD(tree->gtType)))
+    {
+        return false;
+    }
+
+    GenTreeLclVar* lclStore = tree->AsLclVar();
+
+    LCLMasksWeight weight;
+    bool           found = weightsTable->Lookup(lclStore->GetLclNum(), &weight);
+    assert(found);
+
+    if (!weight.ShouldSwitch())
+    {
+        JITDUMP("Local Store V%02d at [%06u] will not be converted. ", lclStore->GetLclNum(), dspTreeID(lclStore));
+        weight.DumpTotalWeight();
+        return false;
+    }
+
+    JITDUMP("Local Store V%02d at [%06u] will be converted. ", lclStore->GetLclNum(), dspTreeID(lclStore));
+    weight.DumpTotalWeight();
+
+    // Update the type of the STORELCL - including the lclvar.
+    lclStore->gtType  = TYP_MASK;
+    LclVarDsc* varDsc = lvaGetDesc(lclStore->GetLclNum());
+    varDsc->lvType    = TYP_MASK;
+
+    if (lclStore->Data()->OperIsConvertMaskToVector())
+    {
+        // Remove the ConvertMaskToVector
+
+        GenTreeHWIntrinsic* convertOp = lclStore->Data()->AsHWIntrinsic();
+        GenTree*            maskOp    = convertOp->Op(1);
+
+        // Remove the convert from the tree.
+        convertOp->gtBashToNOP();
+        lclStore->gtOp1 = maskOp;
+        fgSequenceLocals(stmt);
+
+        JITDUMP("Updated V%02d to store as mask (Removed conversion)\n", lclStore->GetLclNum());
+    }
+    else
+    {
+        // Convert the input of the store to a mask.
+        assert(weight.simdBaseJitType != CORINFO_TYPE_UNDEF);
+        GenTree* convertOp =
+            gtNewSimdCvtVectorToMaskNode(TYP_MASK, lclStore->Data(), weight.simdBaseJitType, weight.simdSize);
+        lclStore->Data() = convertOp;
+
+        JITDUMP("Updated V%02d to store as mask (Added conversion)\n", lclStore->GetLclNum());
+    }
+
+#ifdef DEBUG
+    if (verbose)
+    {
+        gtDispTree(lclStore);
+    }
+#endif
+
+    return true;
+}
+
+//-----------------------------------------------------------------------------
+// fgLCLMasksUpdateLCLVar: For the given lcl var, if the weighting recommends to switch,
+// then update to use the source as a mask.
+//
+// Arguments:
+//     lclVar - The local variable.
+//     stmt - The statement the local vairable is contained in.
+//     weightsTable - table to update.
+//
+void Compiler::fgLCLMasksUpdateLCLVar(GenTreeLclVarCommon* lclVar,
```

Comment:
also instead of `..LclStore` and `...LclVar`, I would just refer to methods for `...CvtMaskToVector` or `...CvtVectorToMask` in the name to easily follow things.

Reply:
Agreed the naming isn't ideal.
With `fgLclMasksUpdateLclVar()` it either removes a `ConvertVectorToMask` or adds a `ConvertMaskToVector`.

A more accurate name would be `fgLclMasksUseLclVarAsMask()`. Not sure how much better that is.

Reply:
I would suggest `fgOptimizeVectorMaskConversions` (similarly `PHASE_LCL_MASKS` -> `PHASE_CONVERT_VECTOR_MASKS`, "Local masks" -> "Optimize vector<->mask conversions")

Reply:
Well, I guess this isn't the entry point of the phase.. for this one I would probably suggest `fgUpdateVectorMaskConversionsInStatement` (but as mentioned in some other feedback I think it doesn't really make sense for this function to take `lclVar` here, and these updates should rather be done directly by a `GenTreeVisitor`)